File: zmq/python/zmq_topic_publisher.py
#!/usr/bin/env python3
"""
ZMQ Publisher with Topics Example
This demonstrates how to publish messages with different topics
"""
import zmq
import time
import random

def main():
    print("=" * 60)
    print("ZMQ TOPIC PUBLISHER")
    print("=" * 60)
    
    context = zmq.Context()
    socket = context.socket(zmq.PUB)
    socket.bind("tcp://*:5556")
    
    print("\n✓ Publisher started on port 5556")
    print("  Publishing messages with 3 topics:")
    print("    - Temperature")
    print("    - Humidity")
    print("    - Pressure")
    print("\nWaiting 2 seconds for subscribers...")
    time.sleep(2)
    print("Starting to publish...\n")
    print("-" * 60)
    
    try:
        message_count = 0
        while True:
            # Randomly choose a topic
            topics = ["Temperature", "Humidity", "Pressure"]
            topic = random.choice(topics)
            
            # Generate data based on topic
            if topic == "Temperature":
                value = random.randint(15, 30)
                unit = "°C"
            elif topic == "Humidity":
                value = random.randint(40, 80)
                unit = "%"
            else:  # Pressure
                value = random.randint(980, 1020)
                unit = "hPa"
            
            # Create message with topic prefix
            message = f"{topic} {value}{unit}"
            
            # Send message (topic is part of the message)
            socket.send_string(message)
            message_count += 1
            
            # Log with color/formatting
            timestamp = time.strftime("%H:%M:%S")
            topic_padded = topic.ljust(12)
            print(f"[{timestamp}] #{message_count:3d} [{topic_padded}] {value}{unit}")
            
            time.sleep(0.5)
            
    except KeyboardInterrupt:
        print("\n" + "-" * 60)
        print(f"\n✓ Published {message_count} messages total")
    
    finally:
        socket.close()
        context.term()
        print("✓ Publisher stopped\n")

if __name__ == "__main__":
    main()

File: zmq/Go/zmq_topic_publisher.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"time"

	zmq "github.com/pebbe/zmq4"
)

func main() {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("GO ZMQ TOPIC PUBLISHER")
	fmt.Println(strings.Repeat("=", 60))

	publisher, _ := zmq.NewSocket(zmq.PUB)
	defer publisher.Close()
	publisher.Bind("tcp://*:5556")

	fmt.Println("\n✓ Publisher started on port 5556")
	time.Sleep(2 * time.Second)
	fmt.Println("Starting to publish...\n")

	topics := []string{"Temperature", "Humidity", "Pressure"}
	msgCount := 0

	for {
		msgCount++
		// Randomly select a topic
		topic := topics[rand.Intn(len(topics))]
		var value int
		var unit string

		switch topic {
		case "Temperature":
			value = rand.Intn(16) + 15
			unit = "°C"
		case "Humidity":
			value = rand.Intn(41) + 40
			unit = "%"
		case "Pressure":
			value = rand.Intn(41) + 980
			unit = "hPa"
		}

		// Construct message: "Topic ValueUnit"
		// e.g., "Temperature 25°C"
		msg := fmt.Sprintf("%s %d%s", topic, value, unit)

		publisher.Send(msg, 0)

		timestamp := time.Now().Format("15:04:05")
		// Formatting output similar to Python's ljust
		fmt.Printf("[%s] #%3d [%-12s] %d%s\n", timestamp, msgCount, topic, value, unit)

		time.Sleep(500 * time.Millisecond)
	}
}

